sane-backends-1.0.25/doc/sane.man
.TH sane 7 "14 Jul 2008" "@PACKAGEVERSION@" "SANE Scanner Access Now Easy"
.IX sane

.SH NAME
sane \- Scanner Access Now Easy: API for accessing scanners

.SH DESCRIPTION
.B SANE
is an application programming interface (API) that provides standardized
access to any raster image scanner hardware. The standardized interface makes
it possible to write just one driver for each scanner device instead of one
driver for each scanner and application.

While
.B SANE
is primarily targeted at a UNIX environment, the standard has been carefully
designed to make it possible to implement the API on virtually any hardware or
operating system.
.PP
This manual page provides a summary of the information available about
.BR SANE .
.PP
If you have trouble getting your scanner detected, read the PROBLEMS section.

.SH TERMINOLOGY

An application that uses the 
.B SANE
interface is called a 
.B SANE frontend.
A driver that implements the 
.B SANE
interface is called a
.B SANE backend.
A 
.B meta backend
provides some means to manage one or more other backends.


.SH "SOFTWARE PACKAGES"
The package 
.RB ` sane\-backends '
contains a lot of backends, documentation (including the 
.B SANE
standard), networking support, and the command line frontend 
.RB ` scanimage '.
The frontends 
.RB ` xscanimage "', `" xcam "', and `" scanadf '
are included in the package
.RB ` sane\-frontends '. 
Both packages can be downloaded from the
.B SANE
homepage 
.RI ( http://www.sane\-project.org/ ).
Information about other frontends and backends can also be found on the
.B SANE
homepage.

.SH "GENERAL INFORMATION"
The following sections provide short descriptions and links to more
information about several aspects of
.B SANE.
A name with a number in parenthesis (e.g. 
.RB ` sane\-dll (5)')
points to a manual page. In this case 
.RB ` "man 5 sane\-dll" '
will display the page. Entries like
.RI ` @DOCDIR@/sane.tex '
are references to text files that were copied to the
.B SANE
documentation directory 
.RI ( @DOCDIR@/ )
during installation. Everything else is a URL to a resource on the web.

.TP 2
.B SANE homepage
Information on all aspects of SANE including a tutorial and a link to the SANE FAQ
can be found on the SANE homepage:
.IR http://www.sane\-project.org/ .
.TP
.B SANE device lists
The 
.B SANE
device lists contain information about the status of 
.B SANE
support for a specific device. If your scanner is not listed there (either
supported or unsupported), please contact us. See section HOW CAN YOU HELP
SANE for details. There are lists for specific releases of SANE, for the
current development version and a search engine:
.IR http://www.sane\-project.org/sane\-supported\-devices.html .
The lists are also installed on your system at @DOCDIR@/.
.TP
.B SANE mailing list
There is a mailing list for the purpose of discussing the SANE standard and its
implementations: sane\-devel.  Despite its name, the list is not only intended
for developers, but also for users. There are also some more lists for special
topics, however, for users, sane\-devel is the right list. How to subscribe and
unsubscribe:
.IR http://www.sane\-project.org/mailing\-lists.html .
.TP
.B SANE IRC channel
The IRC (Internet Relay Chat) channel #sane can be found on the Freenode
network (irc.freenode.net). It's for discussing 
.B SANE
problems, talking about development and general 
.B SANE
related chatting. Before asking for help, please read the other documentation
mentioned in this manual page. The channel's topic is also used for
announcements of problems with SANE infrastructure (mailing lists, web server,
etc.).
.TP
.B Compiling and installing SANE
Look at
.I  @DOCDIR@/README
and the os-dependent README files for information about compiling and
installing
.B SANE.
.TP
.B SCSI configuration
For information about various systems and SCSI controllers see
.BR sane\-scsi (5).
.TP
.B USB configuration
For information about USB configuration see
.BR sane\-usb (5).

.SH "FRONTENDS AND MISCELLANEOUS PROGRAMS"
.TP 2
.B scanimage
Command-line frontend. See 
.BR scanimage (1).
.TP
.B saned
.B SANE
network daemon that allows remote clients to access image acquisition devices
available on the local host. See 
.BR saned (8).
.TP
.B sane\-find\-scanner
Command-line tool to find SCSI and USB scanners and determine their Unix
device files. See
.BR sane\-find\-scanner (1).
.PP
Also, have a look at the
.B sane\-frontends
package (including
.BR xscanimage ", " xcam ", and " scanadf )
and the frontend information page at
.IR http://www.sane\-project.org/sane\-frontends.html .

.SH "BACKENDS FOR SCANNERS"
.TP 2
.B abaton
The SANE backend for Abaton flatbed scanners supports the Scan 300/GS (8bit,
256 levels of gray) and the Scan 300/S (black and white, untested). See
.BR sane\-abaton (5)
for details.
.TP
.B agfafocus
This backend supports AGFA Focus scanners and the Siemens S9036 (untested).
See
.BR sane\-agfafocus (5)
for details.
.TP
.B apple
The SANE backend for Apple flatbed scanners supports the following scanners:
AppleScanner, OneScanner and ColorOneScanner. See
.BR sane\-apple (5)
for details.
.TP
.B artec
The SANE Artec backend supports several Artec/Ultima SCSI flatbed scanners as
well as the BlackWidow BW4800SP and the Plustek 19200S. See
.BR sane\-artec (5)
for details.
.TP
.B artec_eplus48u
The SANE artec_eplus48u backend supports the scanner Artec E+ 48U and re-badged
models like Tevion MD 9693, Medion MD 9693, Medion MD 9705 and Trust Easy
Webscan 19200. See
.BR sane\-artec_eplus48u (5)
for details.
.TP
.B as6e
This is a SANE backend for using the Artec AS6E parallel port interface
scanner. See
.BR sane\-as6e (5)
for details.
.TP
.B avision
This backend supports several Avision based scanners. This includes the
original Avision scanners (like AV 630, AV 620, ...) as well as the HP
ScanJet 53xx and 74xx series, Fujitsu ScanPartner, some Mitsubishi and
Minolta film-scanners.
See 
.BR sane\-avision (5)
for details.
.TP
.B bh
The bh backend provides access to Bell+Howell Copiscan II series document
scanners. See
.BR sane\-bh (5)
for details.
.TP
.B canon
The canon backend supports the CanoScan 300, CanoScan 600, and CanoScan
2700F SCSI flatbed scanners. See
.BR sane\-canon (5)
for details.
.TP
.B canon630u
The canon630u backend supports the CanoScan 630u and 636u USB scanners.  See
.BR sane\-canon630u (5)
for details.
.TP
.B canon_dr
The canon_dr backend supports the Canon DR-Series ADF SCSI and USB scanners. See
.BR sane\-canon_dr (5)
for details.
.TP
.B canon_pp
The canon_pp backend supports the CanoScan FB330P, FB630P, N340P and N640P
parallel port scanners.  See
.BR sane\-canon_pp (5)
for details.
.TP
.B cardscan
This backend provides support for Corex Cardscan USB scanners. See 
.BR sane\-cardscan (5)
for details.
.TP
.B coolscan
This is a SANE backend for Nikon Coolscan film-scanners. See
.BR sane\-coolscan (5)
for details.
.TP
.B coolscan2
This is a SANE backend for Nikon Coolscan film-scanners.
See
.BR sane\-coolscan2 (5)
or
.I http://coolscan2.sourceforge.net
for details.
.TP
.B epjitsu
The epjitsu backend provides support for Epson-based Fujitsu USB scanners. See
.BR sane\-epjitsu (5)
for details.
.TP
.B epson
The SANE epson backend provides support for Epson SCSI, parallel port and USB
flatbed scanners. See
.BR sane\-epson (5)
for details.
.TP
.B fujitsu
The fujitsu backend provides support for most Fujitsu SCSI and USB, flatbed
and adf scanners. See
.BR sane\-fujitsu (5)
for details.
.TP
.B genesys
The genesys backend provides support for several scanners based on the Genesys Logic 
GL646, GL841, GL843, GL847 and GL124 chips like the Medion 6471 and Hewlett-Packard 2300c.
 See
.BR sane\-genesys (5)
for details.
.TP
.B gt68xx
The gt68xx backend provides support for scanners based on the Grandtech
GT-6801 and GT-6816 chips like the Artec Ultima 2000 and several Mustek
BearPaw CU and TA models. Some Genius, Lexmark, Medion, Packard Bell, Plustek,
and Trust scanners are also supported. See
.BR sane\-gt68xx (5)
for details.
.TP
.B hp
The SANE hp backend provides access to Hewlett-Packard ScanJet scanners which
support SCL (Scanner Control Language by HP). See
.BR sane\-hp (5)
for details.
.TP
.B hpsj5s
The SANE backend for the Hewlett-Packard ScanJet 5S scanner. See
.BR sane\-hpsj5s (5)
for details.
.TP
.B hp3500
The SANE backend for the Hewlett-Packard ScanJet 3500 series. See
.BR sane\-hp3500 (5)
for details.
.TP
.B hp3900
The SANE backend for the Hewlett-Packard ScanJet 3900 series. See
.BR sane\-hp3900 (5)
for details.
.TP
.B hp4200
The SANE backend for the Hewlett-Packard ScanJet 4200 series. See
.BR sane\-hp4200 (5)
for details.
.TP
.B hp5400
The SANE backend for the Hewlett-Packard ScanJet 54XXC series. See
.BR sane\-hp5400 (5)
for details.
.TP
.B hpljm1005
The SANE backend for the Hewlett-Packard LaserJet M1005 scanner. See
.BR sane\-hpljm1005(5)
for details.
.TP
.B hs2p
The SANE backend for the Ricoh IS450 family of SCSI scanners. See
.BR sane\-hs2p (5)
for details.
.TP
.B ibm
The SANE backend for some IBM and Ricoh SCSI scanners. See
.BR sane\-ibm (5)
for details.
.TP
.B kodak
The SANE backend for some large Kodak scanners. See
.BR sane\-kodak (5)
for details.
.TP
.B kodakaio
The SANE backend for Kodak AiO printer/scanners. See
.BR sane\-kodakaio (5)
for details.
.TP
.B kvs1025
The SANE backend for Panasonic KV-S102xC scanners. See
.BR sane\-kvs1025 (5)
for details.
.TP
.B leo
This backend supports the Leo S3 and the Across FS-1130, which is a re-badged
LEO FS-1130 scanner. See
.BR sane\-leo (5)
for details.
.TP
.B lexmark
This backend supports the Lexmark X1100 series of USB scanners. See
.BR sane\-lexmark (5)
for details.
.TP
.B ma1509
The ma1509 backend supports the Mustek BearPaw 1200F USB flatbed scanner. See
.BR sane\-ma1509 (5)
for details.
.TP
.B magicolor
The magicolor backend supports the KONICA MINOLTA magicolor 1690MF multi-function printer/scanner/fax. See
.BR sane\-magicolor (5)
for details.
.TP
.B matsushita
This backend supports some Panasonic KVSS high speed scanners. See
.BR sane\-matsushita (5)
for details.
.TP
.B microtek
The microtek backend provides access to the "second generation" Microtek
scanners with SCSI-1 command set. See 
.BR sane\-microtek (5)
for details.
.TP
.B microtek2
The microtek2 backend provides access to some Microtek scanners with a
SCSI-2 command set. See
.BR sane\-microtek2 (5)
for details.
.TP
.B mustek
The SANE mustek backend supports most Mustek SCSI flatbed scanners including the
Paragon and ScanExpress series and the 600 II N and 600 II EP (non-SCSI). Some
Trust scanners are also supported. See
.BR sane\-mustek (5)
for details.
.TP
.B mustek_pp
The mustek_pp backend provides access to Mustek parallel port flatbed
scanners. See
.BR sane\-mustek_pp (5)
for details.
.TP
.B mustek_usb
The mustek_usb backend provides access to some Mustek ScanExpress USB flatbed
scanners. See
.BR sane\-mustek_usb (5)
for details.
.TP
.B mustek_usb2
The mustek_usb2 backend provides access to scanners using the SQ113
chipset like the Mustek BearPaw 2448 TA Pro USB flatbed scanner. See
.BR sane\-mustek_usb2 (5)
for details.
.TP
.B nec
The SANE nec backend supports the NEC PC-IN500/4C SCSI scanner. See
.BR sane\-nec (5)
for details.
.TP
.B niash
The niash backend supports the Agfa Snapscan Touch and the HP ScanJet 3300c,
3400c, and 4300c USB flatbed scanners. See
.BR sane\-niash (5)
for details.
.TP
.B p5
The SANE backend for Primax PagePartner. See
.BR sane\-p5 (5)
for details.
.TP
.B pie
The pie backend provides access to Pacific Image Electronics (PIE) and Devcom
SCSI flatbed scanners. See
.BR sane\-pie (5)
for details.
.TP
.B pixma
The pixma backend supports Canon PIXMA MP series (multi-function devices). See
.BR sane\-pixma (5)
or
.I http://home.arcor.de/wittawat/pixma/
for details.
.TP
.B plustek
The SANE plustek backend supports USB flatbed scanners that use the National
Semiconductor LM983[1/2/3] chipset aka Merlin. Scanners using this LM983x chips
include some models from Plustek, KYE/Genius, Hewlett-Packard, Mustek, Umax,
Epson, and Canon. See
.BR sane\-plustek (5)
for details.
.TP
.B plustek_pp
The SANE plustek_pp backend supports Plustek parallel port flatbed scanners.
Scanners using the Plustek ASIC P96001, P96003, P98001 and P98003 include some
models from Plustek, KYE/Genius, Primax. See
.BR sane\-plustek_pp (5)
for details.
.TP
.B ricoh
The ricoh backend provides access to the following Ricoh flatbed
scanners: IS50 and IS60. See
.BR sane\-ricoh (5)
for details.
.TP
.B s9036
The s9036 backend provides access to Siemens 9036 flatbed scanners. See
.BR sane\-s9036 (5)
for details.
.TP
.B sceptre
The sceptre backend provides access to the Sceptre S1200 flatbed scanner. See
.BR sane\-sceptre (5)
for details.
.TP
.B sharp
The SANE sharp backend supports Sharp SCSI scanners. See
.BR sane\-sharp (5)
for details.
.TP
.B sm3600
The SANE sm3600 backend supports the Microtek ScanMaker 3600 USB scanner. See
.BR sane\-sm3600 (5)
for details.
.TP
.B sm3840
The SANE sm3840 backend supports the Microtek ScanMaker 3840 USB scanner.  See
.BR sane\-sm3840 (5)
for details.
.TP
.B snapscan
The snapscan backend supports AGFA SnapScan flatbed scanners. See
.BR sane\-snapscan (5)
for details.
.TP
.B sp15c
This backend supports the Fujitsu FCPA ScanPartner 15C flatbed scanner. See
.BR sane\-sp15c (5)
for details.
.TP
.B st400
The sane\-st400 backend provides access to Siemens ST400 and ST800. See
.BR sane\-st400 (5)
for details.
.TP
.B tamarack
The SANE tamarack backend supports Tamarack Artiscan flatbed scanners. See
.BR sane\-tamarack (5)
for details.
.TP
.B teco1 teco2 teco3
The SANE teco1, teco2 and teco3 backends support some TECO scanners,
usually sold under the Relisys, Trust, Primax, Piotech, Dextra
names. See
.BR sane\-teco1 "(5), " sane\-teco2 "(5) and " sane\-teco3 (5)
for details.
.TP
.B u12
The sane\-u12 backend provides USB flatbed scanners based on Plustek's ASIC 98003
(parallel-port ASIC) and a GeneSys Logics' USB-parport bridge chip like the
Plustek OpticPro U(T)12. See
.BR sane\-u12 (5)
for details.
.TP
.B umax
The sane\-umax backend provides access to several UMAX-SCSI-scanners and some
Linotype Hell SCSI-scanners. See
.BR sane\-umax (5)
for details.
.TP
.B umax_pp
The sane\-umax_pp backend provides access to Umax parallel port flatbed scanners
and the HP 3200C. See 
.BR sane\-umax_pp (5)
for details.
.TP
.B umax1200u
The sane\-umax1220u backend supports the UMAX Astra 1220U (USB) flatbed scanner
(and also the UMAX Astra 2000U, sort of). See
.BR sane\-umax1220u (5)
for details.
.PP
Also, have a look at the backend information page at
.I http://www.sane\-project.org/sane\-supported\-devices.html
and the list of projects in
.IR @DOCDIR@/PROJECTS .

.SH "BACKENDS FOR DIGITAL CAMERAS"
.TP 2
.B dc210
Backend for Kodak DC210 Digital Camera. See
.BR sane\-dc210 (5).
.TP
.B dc240
Backend for Kodak DC240 Digital Camera. See
.BR sane\-dc240 (5).
.TP
.B dc25
Backend for Kodak DC20/DC25 Digital Cameras. See
.BR sane\-dc25 (5).
.TP
.B dmc
Backend for the Polaroid Digital Microscope Camera. See
.BR sane\-dmc (5).
.TP
.B gphoto2
Backend for digital cameras supported by the gphoto2 library package.  (See
.I http://www.gphoto.org
for more information and a list of supported cameras.)  Gphoto2 supports over
140 different camera models.  However, please note that more development and
testing is needed before all of these cameras will be supported by
.B SANE
backend.  See
.BR sane\-gphoto2 (5).
.TP
.B qcam
Backend for Connectix QuickCam cameras. See 
.BR sane\-qcam (5).
.TP
.B stv680
The sane\-st680 backend provides access to webcams with a stv680 chip. See
.BR sane\-st680 (5)
for details.
.PP
Also, have a look at the backend information page at
.I http://www.sane\-project.org/sane\-supported\-devices.html
and the list of projects in
.IR @DOCDIR@/PROJECTS .

.SH "MISCELLANEOUS BACKENDS"
.TP 2
.B dll
The sane\-dll library implements a
.B SANE
backend that provides access to an arbitrary number of other
.B SANE
backends by dynamic loading. See
.BR sane\-dll (5).
.TP
.B net
The 
.B SANE

sane-backends-1.0.25/japi/Sane.c
/* sane - Scanner Access Now Easy.
   Copyright (C) 1997 Jeffrey S. Freedman
   This file is part of the SANE package.

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2 of the
   License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston,
   MA 02111-1307, USA.

   As a special exception, the authors of SANE give permission for
   additional uses of the libraries contained in this release of SANE.

   The exception is that, if you link a SANE library with other files
   to produce an executable, this does not by itself cause the
   resulting executable to be covered by the GNU General Public
   License.  Your use of that executable is in no way restricted on
   account of linking the SANE library code into it.

   This exception does not, however, invalidate any other reasons why
   the executable file might be covered by the GNU General Public
   License.

   If you submit changes to SANE to the maintainers to be included in
   a subsequent release, you agree by submitting the changes that
   those changes may be distributed with this exception intact.

   If you write modifications of your own for SANE, it is your choice
   whether to permit this exception to apply to your modifications.
   If you do not wish that, delete this exception notice.  */

/**
 **	Sane.c - Native methods for the SANE Java API.
 **
 **	Written: 10/9/97 - JSF
 **/

#include "Sane.h"
#include <sane/sane.h>
#include <string.h>

#include <stdio.h>	/* Debugging */

#ifdef __cplusplus
extern "C" {
#endif
/*
 * Class:     Sane
 * Method:    init
 * Signature: ([I)I
 */
JNIEXPORT jint JNICALL Java_Sane_init
  (JNIEnv *env, jobject jobj, jintArray versionCode)
	{
	jsize len;			/* Gets array length. */
	jint *versionCodeBody;		/* Gets ->array. */
	SANE_Int version;		/* Gets version. */
	SANE_Status status;		/* Get SANE return. */

	status = sane_init(&version, 0);
	len = (*env)->GetArrayLength(env, versionCode);
	versionCodeBody = (*env)->GetIntArrayElements(env, versionCode, 0);
	if (len > 0)			/* Return version. */
		versionCodeBody[0] = version;
	(*env)->ReleaseIntArrayElements(env, versionCode, versionCodeBody, 0);
	return (status);
	}
/*
 * Class:     Sane
 * Method:    exit
 * Signature: ()V
 */
JNIEXPORT void JNICALL Java_Sane_exit
  (JNIEnv *env, jobject jobj)
	{
	sane_exit();
	}


/*
 * Class:     Sane
 * Method:    getDevices
 * Signature: ([LSaneDevice;Z)I
 */
JNIEXPORT jint JNICALL Java_Sane_getDevicesNative
  (JNIEnv *env, jobject jobj, jobjectArray devList, jboolean localOnly)
	{
					/* Gets device list. */
	const SANE_Device **device_list;
	SANE_Status status;		/* Gets status. */
	int devListLen;			/* Gets length of devList. */
	jobject devObj;			/* Gets each SaneDevice object. */
	jclass devClass;		/* Gets SaneDevice class. */
	jfieldID fid;			/* Gets each field ID. */
	int i;

					/* Get the list. */
	status = sane_get_devices(&device_list, localOnly);	
	if (status != SANE_STATUS_GOOD)
		return (status);
					/* Get length of Java array. */
	devListLen = (*env)->GetArrayLength(env, devList);
					/* Return devices to user. */
	for (i = 0; i < devListLen - 1 && device_list[i]; i++)
		{
					/* Get Java object, class. */
		devObj = (*env)->GetObjectArrayElement(env, devList, i);
		devClass = (*env)->GetObjectClass(env, devObj);
					/* Fill in each member. */
		fid = (*env)->GetFieldID(env, devClass, "name",
							"Ljava/lang/String;");
		(*env)->SetObjectField(env, devObj, fid,
			(*env)->NewStringUTF(env, device_list[i]->name));
		fid = (*env)->GetFieldID(env, devClass, "vendor",
							"Ljava/lang/String;");
		(*env)->SetObjectField(env, devObj, fid,
			(*env)->NewStringUTF(env, device_list[i]->vendor));
		fid = (*env)->GetFieldID(env, devClass, "model",
							"Ljava/lang/String;");
		(*env)->SetObjectField(env, devObj, fid,
			(*env)->NewStringUTF(env, device_list[i]->model));
		fid = (*env)->GetFieldID(env, devClass, "type",
							"Ljava/lang/String;");
		(*env)->SetObjectField(env, devObj, fid,
			(*env)->NewStringUTF(env, device_list[i]->type));
		}
					/* End list with a null. */
	(*env)->SetObjectArrayElement(env, devList, i, 0);
	return (status);
	}

/*
 * Class:     Sane
 * Method:    open
 * Signature: (Ljava/lang/String;[J)I
 */
JNIEXPORT jint JNICALL Java_Sane_open
  (JNIEnv *env, jobject jobj, jstring deviceName, jintArray handle)
	{
	SANE_Handle sane_handle;	/* Gets handle. */
	jint s_handle;
	const char *device_name;	/* Gets dev. name. */
	int status;			/* Gets return code. */

	device_name = (*env)->GetStringUTFChars(env, deviceName, 0);
					/* Open it. */
	status = sane_open(device_name, &sane_handle);
	(*env)->ReleaseStringUTFChars(env, deviceName, device_name);
					/* Return handle. */
	s_handle = (jint) sane_handle;
	(*env)->SetIntArrayRegion(env, handle, 0, 1, &s_handle);
	return (status);
	}

/*
 * Class:     Sane
 * Method:    close
 * Signature: (J)V
 */
JNIEXPORT void JNICALL Java_Sane_close
  (JNIEnv *env, jobject jobj, jint handle)
	{
	sane_close((SANE_Handle) handle);
	}

/*
 * Class:     Sane
 * Method:    getOptionNative
 * Signature: (IILSaneOption;)V
 */
JNIEXPORT void JNICALL Java_Sane_getOptionNative
  (JNIEnv *env, jobject jobj, jint handle, jint option, jobject optObj)
	{
	jclass optClass;		/* Gets its class. */
	jfieldID fid;			/* Gets each field ID. */
	jstring str;			/* Gets strings. */

					/* Get info from sane. */
	const SANE_Option_Descriptor *sopt = sane_get_option_descriptor(
			(SANE_Handle) handle, option);
					/* Get class info. */
	optClass = (*env)->GetObjectClass(env, optObj);
					/* Fill in each member. */
	fid = (*env)->GetFieldID(env, optClass, "name", "Ljava/lang/String;");
	if (!sopt)			/* Failed. */
		{			/* Set name to null. */
		(*env)->SetObjectField(env, optObj, fid, 0);
		return;
		}
					/* Return name. */
	(*env)->SetObjectField(env, optObj, fid,
				(*env)->NewStringUTF(env, sopt->name));
					/* Return title. */
	fid = (*env)->GetFieldID(env, optClass, "title", "Ljava/lang/String;");
	str = sopt->title ? (*env)->NewStringUTF(env, sopt->title) : 0;
	(*env)->SetObjectField(env, optObj, fid, str);
					/* Return descr. */
	fid = (*env)->GetFieldID(env, optClass, "desc", "Ljava/lang/String;");
	(*env)->SetObjectField(env, optObj, fid,
				(*env)->NewStringUTF(env, sopt->desc));
					/* Return type. */
	fid = (*env)->GetFieldID(env, optClass, "type", "I");
	(*env)->SetIntField(env, optObj, fid, sopt->type);
					/* Return unit. */
	fid = (*env)->GetFieldID(env, optClass, "unit", "I");
	(*env)->SetIntField(env, optObj, fid, sopt->unit);
					/* Return size. */
	fid = (*env)->GetFieldID(env, optClass, "size", "I");
	(*env)->SetIntField(env, optObj, fid, sopt->size);
					/* Return cap. */
	fid = (*env)->GetFieldID(env, optClass, "cap", "I");
	(*env)->SetIntField(env, optObj, fid, sopt->cap);
					/* Return constraint_type. */
	fid = (*env)->GetFieldID(env, optClass, "constraintType", "I");
	(*env)->SetIntField(env, optObj, fid, sopt->constraint_type);
	/*
	 *	Now for the constraint itself.
	 */
	if (sopt->constraint_type == SANE_CONSTRAINT_RANGE)
		{
					/* Create range object. */
		jclass rangeClass = (*env)->FindClass(env, "SaneRange");
		jobject range = (*env)->AllocObject(env, rangeClass);
					/* Fill in fields. */
		fid = (*env)->GetFieldID(env, rangeClass, "min", "I");
		(*env)->SetIntField(env, range, fid, 
						sopt->constraint.range->min);
		fid = (*env)->GetFieldID(env, rangeClass, "max", "I");
		(*env)->SetIntField(env, range, fid, 
						sopt->constraint.range->max);
		fid = (*env)->GetFieldID(env, rangeClass, "quant", "I");
		(*env)->SetIntField(env, range, fid, 
						sopt->constraint.range->quant);
		fid = (*env)->GetFieldID(env, optClass, "rangeConstraint",
						"LSaneRange;");
					/* Store range. */
		(*env)->SetObjectField(env, optObj, fid, range);
		}
	else if (sopt->constraint_type == SANE_CONSTRAINT_WORD_LIST)
		{			/* Get array of integers. */
		jintArray wordList;
		jint *elements;
		int i;
					/* First word. is the length.	*/
		wordList = (*env)->NewIntArray(env, 
				sopt->constraint.word_list[0]);
					/* Copy in the integers.	*/
		elements = (*env)->GetIntArrayElements(env, wordList, 0);
		for (i = 0; i < sopt->constraint.word_list[0]; i++)
			elements[i] = sopt->constraint.word_list[i];
		(*env)->ReleaseIntArrayElements(env, wordList, elements, 0);
					/* Set the field. */
		fid = (*env)->GetFieldID(env, optClass, "wordListConstraint",
									"[I");
		(*env)->SetObjectField(env, optObj, fid, wordList);
		}
	else if (sopt->constraint_type == SANE_CONSTRAINT_STRING_LIST)
		{
		jclass stringClass = (*env)->FindClass(env, "java/lang/String");
		jobjectArray stringList;
		int len;		/* Gets # elements */
		int i;

		for (len = 0; sopt->constraint.string_list[len]; len++)
			;
		stringList = (*env)->NewObjectArray(env, len + 1, 
							stringClass, 0);
					/* Add each string. */
		for (i = 0; i < len; i++)
			{
			(*env)->SetObjectArrayElement(env, stringList, i,
				(*env)->NewStringUTF(env,
					sopt->constraint.string_list[i]));
			}
					/* 0 at end. */
		(*env)->SetObjectArrayElement(env, stringList, len, 0);
					/* Set the field. */
		fid = (*env)->GetFieldID(env, optClass, 
			"stringListConstraint", "[Ljava/lang/String;");
		(*env)->SetObjectField(env, optObj, fid, stringList);
		}
	}

/*
 * Class:     Sane
 * Method:    getControlOption
 * Signature: (II[I[I)I
 */
JNIEXPORT jint JNICALL Java_Sane_getControlOption__II_3I_3I
  (JNIEnv *env, jobject jobj, jint handle, jint option, jintArray value, 
						jintArray info)
	{
	SANE_Status status;		/* Gets status. */
	SANE_Int i;			/* Gets info. passed back. */
	int v;

	status = sane_control_option((SANE_Handle) handle, option,
			SANE_ACTION_GET_VALUE, &v, &i);
	if (value)
		(*env)->SetIntArrayRegion(env, value, 0, 1, &v);
	if (info)
		(*env)->SetIntArrayRegion(env, info, 0, 1, &i);
	return (status);
	}

/*
 * Class:     Sane
 * Method:    getControlOption
 * Signature: (II[B[I)I
 */
JNIEXPORT jint JNICALL Java_Sane_getControlOption__II_3B_3I
  (JNIEnv *env, jobject jobj, jint handle, jint option, jbyteArray value, 
						jintArray info)
	{
	SANE_Status status;		/* Gets status. */
	SANE_Int i;			/* Gets info. passed back. */
	char *str;

	str = (*env)->GetByteArrayElements(env, value, 0);
	status = sane_control_option((SANE_Handle) handle, option,
			SANE_ACTION_GET_VALUE, str, &i);
	(*env)->ReleaseByteArrayElements(env, value, str, 0);
	if (info)
		(*env)->SetIntArrayRegion(env, info, 0, 1, &i);
	return (status);
	}

/*
 * Class:     Sane
 * Method:    setControlOption
 * Signature: (IIII[I)I
 */
JNIEXPORT jint JNICALL Java_Sane_setControlOption__IIII_3I
  (JNIEnv *env, jobject jobj, jint handle, jint option, jint action, 
				jint value, jintArray info)
	{
	SANE_Status status;		/* Gets status. */
	SANE_Int i;			/* Gets info. passed back. */
	status = sane_control_option((SANE_Handle) handle, option, action,
					&value, &i);
	if (info)
		(*env)->SetIntArrayRegion(env, info, 0, 1, &i);
	return (status);
	}

/*
 *	Get string length.  This exists because sometimes strings seem to be
 *	padded with negatives.
 */

static int String_length
	(
	const char *str
	)
	{
	const char *ptr;
	for (ptr = str; *ptr > 0; ptr++)
		;
	return ((int) (ptr - str));
	}

/*
 * Class:     Sane
 * Method:    setControlOption
 * Signature: (IIILjava/lang/String;[I)I
 */
JNIEXPORT jint JNICALL Java_Sane_setControlOption__IIILjava_lang_String_2_3I
  (JNIEnv *env, jobject jobj, jint handle, jint option, jint action,
				jstring value, jintArray info)
	{
	SANE_Status status;		/* Gets status. */
	SANE_Int i;			/* Gets info. passed back. */
	const char *valuep;
	char buf[512];			/* Hope this is big enough. */
	int len;			/* Gets string length. */

	valuep = (*env)->GetStringUTFChars(env, value, 0);
	len = String_length(valuep);
	if (len >= sizeof(buf))
		len = sizeof(buf) - 1;
	strncpy(buf, valuep, len);
	buf[len] = 0;			/* Insure it's 0-delimited. */
	status = sane_control_option((SANE_Handle) handle, option, action,
					(void *) &buf[0], &i);
					/* +++++++Want to return new val? */
	(*env)->ReleaseStringUTFChars(env, value, valuep);
	if (info)
		(*env)->SetIntArrayRegion(env, info, 0, 1, &i);
	return (status);
	}

/*
 * Class:     Sane
 * Method:    getParameters
 * Signature: (ILSaneParameters;)I
 */
JNIEXPORT jint JNICALL Java_Sane_getParameters
  (JNIEnv *env, jobject jobj, jint handle, jobject paramsObj)
	{
	SANE_Status status;		/* Gets status.	*/
	SANE_Parameters params;		/* Gets params. */
	jclass paramsClass;		/* Gets its class. */
	jfieldID fid;			/* Gets each field ID. */

	status = sane_get_parameters((SANE_Handle) handle, &params);
					/* Get class info. */
	paramsClass = (*env)->GetObjectClass(env, paramsObj);
					/* Fill in each member. */
	fid = (*env)->GetFieldID(env, paramsClass, "format", "I");
	(*env)->SetIntField(env, paramsObj, fid, params.format);
	fid = (*env)->GetFieldID(env, paramsClass, "lastFrame", "Z");
	(*env)->SetBooleanField(env, paramsObj, fid, params.last_frame);
	fid = (*env)->GetFieldID(env, paramsClass, "bytesPerLine", "I");
	(*env)->SetIntField(env, paramsObj, fid, params.bytes_per_line);
	fid = (*env)->GetFieldID(env, paramsClass, "pixelsPerLine", "I");
	(*env)->SetIntField(env, paramsObj, fid, params.pixels_per_line);
	fid = (*env)->GetFieldID(env, paramsClass, "lines", "I");
	(*env)->SetIntField(env, paramsObj, fid, params.lines);
	fid = (*env)->GetFieldID(env, paramsClass, "depth", "I");
	(*env)->SetIntField(env, paramsObj, fid, params.depth);
	return (status);
	}

/*
 * Class:     Sane
 * Method:    start
 * Signature: (I)I
 */
JNIEXPORT jint JNICALL Java_Sane_start
  (JNIEnv *env, jobject jobj, jint handle)
	{
	return (sane_start((SANE_Handle) handle));
	}

/*
 * Class:     Sane
 * Method:    read
 * Signature: (I[BI[I)I
 */
JNIEXPORT jint JNICALL Java_Sane_read
  (JNIEnv *env, jobject jobj, jint handle, jbyteArray data, jint maxLength,
					jintArray length)
	{
	int status;
	jbyte *dataElements;
	int read_len;			/* # bytes read. */

					/* Get actual data ptr. */
	dataElements = (*env)->GetByteArrayElements(env, data, 0);
					/* Do the read. */
	status = sane_read((SANE_Handle) handle, dataElements, 
						maxLength, &read_len);
	(*env)->ReleaseByteArrayElements(env, data, dataElements, 0);
					/* Return # bytes read. */
	(*env)->SetIntArrayRegion(env, length, 0, 1, &read_len);
	return (status);
	}

/*
 * Class:     Sane
 * Method:    cancel
 * Signature: (I)V
 */
JNIEXPORT void JNICALL Java_Sane_cancel
  (JNIEnv *env, jobject jobj, jint handle)
	{
	sane_cancel((SANE_Handle) handle);
	}

/*
 * Class:     Sane
 * Method:    strstatus
 * Signature: (I)Ljava/lang/String;
 */
JNIEXPORT jstring JNICALL Java_Sane_strstatus
  (JNIEnv *env, jobject jobj, jint status)
	{
	const char *str = sane_strstatus(status);
	return ((*env)->NewStringUTF(env, str));
	}

#ifdef __cplusplus
}
#endif
